addons/l10n_it_edi_withholding/models/account_move.py

```diff
@@ -186,21 +189,57 @@ def _l10n_it_edi_get_extra_info(self, company, document_type, body_tree, incomin
                  + type_tax_use_domain),
                 vat_only=False)
             if pension_fund_tax:
-                pension_fund_taxes.append(pension_fund_tax)
+                pension_fund_taxes[vat_tax_factor_percent] = pension_fund_tax
             else:
                 message_to_log.append(Markup("%s<br/>%s") % (
                     _("Pension Fund tax not found"),
                     self.env['account.move']._compose_info_message(body_tree, '.'),
                 ))
         extra_info["pension_fund_taxes"] = pension_fund_taxes
 
+        # If the AssoSoftware specs are used on the invoice, then only apply
+        # the Pension Fund tax to the lines that show an AswCassPre
+        # additional tag (AltriDatiGestionali)
+        selector = ".//AltriDatiGestionali/TipoDato[contains(text(), 'AswCassPre')]"
+        if self.get_tag(body_tree, selector) is not None:
+            extra_info["pension_fund_assosoftware_tags"] = True
+
         return extra_info, message_to_log
 
+    def get_tag(self, element, selector):
+        if element is None:
+            return None
+        sub = element.xpath(selector)
+        return sub[0] if (sub is not None and sub) else None
+
+    def _get_pension_fund_tax_for_line(self, element, extra_info):
+        """ If there are AssoSoftware specific AltriDatiGestionale 'AswCassPre'
+            tags that specify which lines have pension funds, use them.
+            Otherwise apply the pension fund on all lines that have the related AliquotaIVA
+        """
+        pension_fund_map = extra_info.get('pension_fund_taxes', {})
+        if pension_fund_map and extra_info.get('pension_fund_assosoftware_tags'):
+            selector = ".//AltriDatiGestionali[TipoDato[contains(text(),'AswCassPre')]]/RiferimentoTesto"
+            if (reference_tag := self.get_tag(element, selector)) is not None:
+                pension_fund_taxes = pension_fund_map.values()
+                if match := re.match(r"(?P<kind>TC\d\d) \((?P<tax_rate>\d+)%\)", reference_tag.text):
+                    if pension_fund_tax := next((
+                        x for x in pension_fund_taxes
+                        if x.l10n_it_pension_fund_type == match['kind']
+                        and x.amount == float(match['tax_rate'])
+                    ), None):
+                        return pension_fund_tax
+        elif (tax_rate_tag := self.get_tag(element, './/AliquotaIVA')) is not None:
+            tax_rate = float(tax_rate_tag.text)
+            if pension_fund_tax := pension_fund_map.get(tax_rate):
+                return pension_fund_tax
+
     def _l10n_it_edi_import_line(self, element, move_line_form, extra_info=None):
         messages_to_log = super()._l10n_it_edi_import_line(element, move_line_form, extra_info)
 
         type_tax_use_domain = extra_info['type_tax_use_domain']
```

Comment:
The default value for `extra_info` is `None` :eyes:

Reply:
This is a sad story of missed refactoring.
It started as a "bugfix" in stable (all this stuff were features born on stable ofc. Italians like to stay 2 versions behind either on premise or on `Odoo.sh`).
There was no parameter. We added one that has a default `= None` to the reading of the header, to pass some info to the line-reading function. This parameter is only used by this module (withholding), and it should always fill it up.
But ... the default is still None.
And I still want to refactor all this in `master`, never found the time.

Do you have any idea for a quick fix that doesn't suck?

Reply:
I don't know if `extra_info = extra_info or {}` would help?
It would at least get rid of a potential traceback...

Reply:
Well, then it's already done in the `super()` call on the first line of the function.
I will repeat it but I don't see it being particularly useful, unless we change the chain of inheritance. Frankly, in these cases, I'd prefer a traceback.

Reply:
> Well, then it's already done in the super() call on the first line of the function.

Doesn't prevent from calling `_l10n_it_edi_import_line` without the parameter. The fact that the super call uses `{}` internally doesn't change the behavior in this extension...